server/tests/raft_metadata_cluster.rs

```diff
@@ -0,0 +1,265 @@
+    };
+    let client = create_client(metadata_store_client_options)
+        .await
+        .expect("to not fail");
+
+    let value = Value::new(42);
+    let value_version = value.version();
+    let key = ByteString::from_static("my-key");
+    client
+        .put(key.clone(), &value, Precondition::DoesNotExist)
+        .await?;
+
+    let stored_value = client.get::<Value>(key.clone()).await?;
+    assert_eq!(stored_value, Some(value));
+
+    let stored_value_version = client.get_version(key.clone()).await?;
+    assert_eq!(stored_value_version, Some(value_version));
+
+    let new_value = Value::new(1337);
+    let new_value_version = new_value.version();
+    let failed_precondition = client
+        .put(
+            key.clone(),
+            &new_value,
+            Precondition::MatchesVersion(value_version.next()),
+        )
+        .await;
+    assert_that!(
+        failed_precondition,
+        err(pat!(WriteError::FailedPrecondition(_)))
+    );
+
+    client
+        .put(
+            key.clone(),
+            &new_value,
+            Precondition::MatchesVersion(value_version),
+        )
+        .await?;
+    let stored_new_value = client.get::<Value>(key.clone()).await?;
+    assert_eq!(stored_new_value, Some(new_value));
+
+    client
+        .delete(key.clone(), Precondition::MatchesVersion(new_value_version))
+        .await?;
+    assert!(client.get::<Value>(key.clone()).await?.is_none());
+
+    cluster.graceful_shutdown(Duration::from_secs(3)).await?;
+
+    Ok(())
+}
+
+#[test_log::test(restate_core::test)]
+async fn raft_metadata_cluster_chaos_test() -> googletest::Result<()> {
+    let num_nodes = 3;
+    let chaos_duration = Duration::from_secs(20);
+    let mut base_config = Configuration::default();
+    base_config.metadata_server.kind = MetadataServerKind::Raft(RaftOptions {
+        raft_election_tick: 5,
+        raft_heartbeat_tick: 2,
+        ..RaftOptions::default()
+    });
+
+    let nodes = Node::new_test_nodes(
+        base_config,
+        BinarySource::CargoTest,
+        EnumSet::empty(),
+        num_nodes,
+        true,
+    );
+    let mut cluster = Cluster::builder()
+        .cluster_name("raft_metadata_cluster_smoke_test")
+        .nodes(nodes)
+        .temp_base_dir()
+        .build()
+        .start()
+        .await?;
+
+    cluster.wait_healthy(Duration::from_secs(30)).await?;
+
+    let addresses = cluster
+        .nodes
+        .iter()
+        .map(|node| node.node_address().clone())
+        .collect();
+
+    let metadata_store_client_options = MetadataStoreClientOptions {
+        metadata_store_client: MetadataStoreClient::Embedded { addresses },
+        ..MetadataStoreClientOptions::default()
+    };
+    let client = create_client(metadata_store_client_options)
+        .await
+        .expect("to not fail");
+
+    let start_chaos = Instant::now();
+
+    let chaos_handle = TaskCenter::spawn_unmanaged(TaskKind::Background, "chaos", async move {
```

Comment:
If only getting a handle on chaos was this easy in real life :-)

Reply:
That would make for an awesome business ;-)